File: 503-next-greater-element-ii/next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> result(n, -1);
        stack<int> stk;  // stores indices

        for (int i = 0; i < 2 * n; i++) {
            int idx = i % n;

            while (!stk.empty() && nums[stk.top()] < nums[idx]) {
                result[stk.top()] = nums[idx];
                stk.pop();
            }

            if (i < n) {
                stk.push(idx);
            }
        }

        return result;
    }
};

File: 503-next-greater-element-ii/next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: list[int]) -> list[int]:
        n = len(nums)
        result = [-1] * n
        stack = []  # stores indices, monotonic decreasing values

        # Iterate 2n times to simulate circular traversal
        for i in range(2 * n):
            idx = i % n

            while stack and nums[stack[-1]] < nums[idx]:
                result[stack.pop()] = nums[idx]

            # Only push indices during the first pass to avoid duplicate processing
            if i < n:
                stack.append(idx)

        return result
        